server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -285,29 +290,43 @@ public TemporalAccessor parse(String input) {
      * @throws DateTimeParseException when unable to parse with any parsers
      */
     private TemporalAccessor doParse(String input) {
-        if (parsers.size() > 1) {
-            Object object = null;
-            if (canCacheLastParsedFormatter && lastParsedformatter != null) {
-                ParsePosition pos = new ParsePosition(0);
-                object = lastParsedformatter.toFormat().parseObject(input, pos);
-                if (parsingSucceeded(object, input, pos)) {
-                    return (TemporalAccessor) object;
-                }
+        final ParsePosition pos = new ParsePosition(0);
+        //Copy to local from volatile to avoid error prune in multi-threads,
+        //other threads may update lastParsedformatter
+        final DateTimeFormatter cachedFormatter = lastParsedformatter;
+        Object object = null;
+        if (canCacheLastParsedFormatter && cachedFormatter != null) {
+            object = tryParseObject(cachedFormatter, input, pos);
+            if (parsingSucceeded(object, input, pos)) {
+                return (TemporalAccessor) object;
             }
-            for (DateTimeFormatter formatter : parsers) {
-                ParsePosition pos = new ParsePosition(0);
-                object = formatter.toFormat().parseObject(input, pos);
+        }
+        for (DateTimeFormatter formatter : parsers) {
+            if(canCacheLastParsedFormatter && formatter != cachedFormatter) {
+                object = tryParseObject(formatter, input, pos);
                 if (parsingSucceeded(object, input, pos)) {
                     lastParsedformatter = formatter;
                     return (TemporalAccessor) object;
                 }
             }
+        }
 
-            throw new DateTimeParseException("Failed to parse with all enclosed parsers", input, 0);
+        throw new DateTimeParseException("Failed to parse with all enclosed parsers", input, 0);
+    }
+
+    private Object tryParseObject(final DateTimeFormatter formatter, final String input, final ParsePosition pos) {
+        //Reset parse position
+        pos.setIndex(0);
+        pos.setErrorIndex(-1);
+        try {
+            return formatter.toFormat().parseObject(input, pos);
+        } catch (final DateTimeException ex) {
```

Comment:
The bug is caused by DateTimeException introduced by JSR-310, the original code doesn't catch  DateTimeException causing the failure in data time detection during indexing.